### ISSUE TITLE
General account payment ArTran written specifically for installment contract purposes

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#1101]
_Imported from trac ticket 1101.
Created by **kbower**
Opened in trac: Sep 28, 2011, 8:36:53 AM
Last modified in trac: Sep 30, 2011, 4:18:10 PM_
***
_by kbower:_
The `ArTran` method `_record_account_payment()` needs to be reworked for when payments are being made for customer account.  It was written specifically for installment contracts.
```python
    def _record_account_payment(self, tran_codeid):
        # paymenttypeid is not nec. required for ArTran, but it is for payment
        if self.required_check('paymenttypeid'):
            if self.writes_artran():
                self.status = u'P'
                self.transactiondefaults(tran_codeid, None, self.sourcesiteid, self.customerid)
                #self.set_defaults() # currently does nothing
                # validate() with check required
                self.validate()
                self.artype = u'I'
                if not getattr(DBSession(),'errors_exist',False):
                    DBSession.add(self)
``` * for account payments, self.artype = u'I' is wrong
 * my guess is self.status = u'P' is wrong
 * `if self.writes_artran()` needs an else statement.  In other words, if the type of Artran isn't self.writes_artran(), we can't just silently ignore the transaction and make the user think they did something.  We should raise an APIError
 * make sure validate() function is correct for this case
 * compare all columns from a GERS generated record to make sure we have it correct.
 * Also, I don't understand why we only allow DEPOSIT instead of also allowing PAYMENT.

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 28, 2011, 8:54:59 AM:_

Currently, when you use the customer account payment functionality (eventual replacement for GERS POA), we are writing an Ar_Trn with AR_TP = 'I' and STAT_CD = 'P', but we think that is only appropriate for installment contract payments.

Our questions are:
* When should the AR_TP be 'I'?   I see Kane's has some from PSOE of this type, but I thought it only had to do with installment contracts
* When should the STAT_CD be 'P'?
* When a customer account receives a payment (POA), what field should be set in which ways?  We will need an example on a TRAIN system so we can get this correct and compare each column to what we write.

If you have used this functionality with LIVE (El Dorado has some AR_TRNs with AR_TP and ORIGIN_CD = 'APP', which is our application), then we may have mixed up your accounting if there is some processing that needs to take place that moves STAT_CD to 'P'.  If that is the case, then please investigate this and sorry for the inconvenience.

I would much appreciate it if both companies would reply so we can get agreement on how this should be working.

Thank you very much,
Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 28, 2011, 12:28:05 PM:_

Kent,

1.        'I' is only for installment contracts as you noted and the AR_TP
should be 'O'

2.        The STAT_CD should be 'T'

3.        Take a look at customer COSST51800 with a payment on 07/15/11 on
the train database for an example.

Thanks,

Wayne Liburd

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 28, 2011, 12:45:28 PM:_

I am very suspicious about 
```python
self.status = u'P'
```for installment payments as well.  Doesn't some accounting report or interface process these transactions and then mark them u'P'?  This almost seems like writing an ITL and marking it uploaded = 'Y' to me....  We sure that should be P for installment contracts?

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 28, 2011, 12:49:24 PM:_

By the way, why are there a couple ar_trn records in your live system marked with AR_TP = 'I'?  Do you do installment contracts?

Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 28, 2011, 1:23:55 PM:_

Kent,

Those transactions were not payments however most likely input errors.

Wayne

**Comment by kbower** [comment]
_**sshah** trac comment on Sep 30, 2011, 11:50:34 AM:_

(Changeset [[1763]](../commit/7a49d4bafed511cb45dc1907dfaa790cbdb8517b))
Fixes #1101 record correct artype and status when accepting payment for order vs. installment

**Comment by kbower** [comment]
_**sshah** trac status change on Sep 30, 2011, 11:50:34 AM:_ **closed**

**Comment by kbower** [comment]
_**sshah** trac comment on Sep 30, 2011, 4:18:10 PM:_

(Changeset [[1765]](../commit/512b2db43db92f1ae9a15ba14ee3ec4d894aa049))
Fixes #1101 payment can be made for all the different type of transaction types.
